### PR TITLE
add failure policy and pause

### PR DESCRIPTION
**Pull request by xlgao-zju**
### Ⅰ. Describe what this PR does
- Add pause, so we can pause the job and we can also remouse it.
- Add failure policy to define the behavior, when a failed pod is found.

> Note: `BackoffLimit` is removed, cause we add `RestartLimit` in `FailurePolicy`.

### Ⅱ. Does this pull request fix one issue?

"NONE" 

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
UT list:
- TestJobFailurePolicyTypeContinue
- TestJobFailurePolicyTypeFailFast
- TestJobFailurePolicyPause
- TestJobSetPaused

### Ⅴ. Special notes for reviews

Signed-off-by: Xianglin Gao <xianglin.gxl@alibaba-inc.com>

### CODE REVIEW

File: pkg/controller/broadcastjob/broadcastjob_controller.go [hunk 1]
```diff
@@ -188,24 +192,50 @@ func (r *ReconcileBroadcastJob) Reconcile(request reconcile.Request) (reconcile.
 	failed := int32(len(failedPods))
 	succeeded := int32(len(succeededPods))
 
-	oldNumConditions := len(job.Status.Conditions)
+	//oldNumConditions := len(job.Status.Conditions)
```
Comment (Fei-Guo) [inline pkg/controller/broadcastjob/broadcastjob_controller.go]: Remove this line?

File: pkg/apis/apps/v1alpha1/broadcastjob_types.go [hunk 2]
```diff
@@ -122,8 +130,43 @@ type BroadcastJobStatus struct {
...
+	// +optional
+	Phase BroadcastJobPhase `json:"phase" protobuf:"varint,8,opt,name=phase"`
 }
 
+// BroadcastJobPhase indicates the phase of the job.
+type BroadcastJobPhase string
+
+const (
+	// PhaseCompleted means the job is completed.
+	PhaseCompleted BroadcastJobPhase = "completed"
```
Comment (Fei-Guo) [inline pkg/apis/apps/v1alpha1/broadcastjob_types.go]: When will controller reach this phase? I don't see the use of this.

File: pkg/controller/broadcastjob/broadcastjob_controller.go [hunk 3]
```diff
@@ -188,24 +192,50 @@ func (r *ReconcileBroadcastJob) Reconcile(request reconcile.Request) (reconcile.
...
-		if err != nil {
-			klog.Errorf("failed to deleteJobPods for job %s,", job.Name)
+		switch job.Spec.FailurePolicy {
+		case appsv1alpha1.FailurePolicyPause:
+			r.recorder.Event(job, corev1.EventTypeWarning, "Paused", "job is paused, due to failed pod")
+			job.Spec.Paused = true
+			job.Status.Phase = appsv1alpha1.PhasePaused
+			return reconcile.Result{RequeueAfter: requeueAfter}, r.updateJobStatus(request, job)
+		case appsv1alpha1.FailurePolicyFastFailed:
+			// Handle Job failures, delete all active pods
```
Comment (Fei-Guo) [inline pkg/controller/broadcastjob/broadcastjob_controller.go]: Should change phase to failed here?

File: pkg/controller/broadcastjob/broadcastjob_controller.go [hunk 4]
```diff
@@ -188,24 +192,50 @@ func (r *ReconcileBroadcastJob) Reconcile(request reconcile.Request) (reconcile.
...
+			failed, active, err = r.deleteJobPods(job, activePods, failed, active)
+			if err != nil {
+				klog.Errorf("failed to deleteJobPods for job %s,", job.Name)
+			}
+			requeueAfter = finishJob(job, appsv1alpha1.JobFailed, failureMessage)
+			r.recorder.Event(job, corev1.EventTypeWarning, failureReason,
+				fmt.Sprintf("%s: %d pods succeeded, %d pods failed", failureMessage, succeeded, failed))
+		case appsv1alpha1.FailurePolicyContinue:
```
Comment (Fei-Guo) [inline pkg/controller/broadcastjob/broadcastjob_controller.go]: Job can fail due to pod failure or hitting ActiveDeadline. Should we continue in the latter case?

Also, it seems that the Continue logic does not work, the reconcilePods method is not called in this code path.

**Comment by Fei-Guo** [comment]
/lgtm
/assign @jian-he 
Assign to Jian for final approve.

File: pkg/apis/apps/v1alpha1/broadcastjob_types.go [hunk 5]
```diff
@@ -123,8 +131,43 @@ type BroadcastJobStatus struct {
...
+
+// FailurePolicy indicates the behavior of the job, when failed pod is found.
+type FailurePolicy string
+
+const (
+	// FailurePolicyContinue means the job will be still running, when failed pod is found.
+	FailurePolicyContinue FailurePolicy = "Continue"
+
+	// FailurePolicyFailFast means the job will be failed, when failed pod is found.
+	FailurePolicyFailFast FailurePolicy = "FailFast"
```
Comment (jian-he) [inline pkg/apis/apps/v1alpha1/broadcastjob_types.go]: API 语义跟 CompletionPolicy 有些重复，要不要merge到completion policy里面。具体的：
FailFast ==“ CompletionPolicy 中的 BackoffLimit = 0”
Continue == “CompletionPolicy 中的BackoffLimit = 无穷”

**Comment by jian-he** [comment]
@xlgao-zju  Could you clarify the difference for `pause` and `continue` ?

Does ` CompletionPolicy = Never` equal to `FailurePolicy = Continue`  ?

**Comment by xlgao-zju** [comment]
@hejian
PR is updated as we discussed offline. And UT is added.

File: pkg/apis/apps/v1alpha1/defaults.go [hunk 6]
```diff
@@ -36,4 +36,8 @@ func SetDefaultsBroadcastJob(job *BroadcastJob) {
 		parallelismIntStr := intstr.FromInt(int(parallelism))
 		job.Spec.Parallelism = &parallelismIntStr
 	}
+
+	if job.Spec.FailurePolicy.Type == "" {
```
Comment (Fei-Guo) [inline pkg/apis/apps/v1alpha1/defaults.go]: SetDefaultBroadcastJob func should be removed from this file (can you do this in this change? Thanks)

Please move the initialization code to 

https://github.com/openkruise/kruise/blob/e0e67c8faaf3d5cd5afc50028374caf6af859460/pkg/webhook/default_server/broadcastjob/mutating/broadcastjob_create_update_handler.go#L57

File: pkg/controller/broadcastjob/broadcastjob_controller.go [hunk 7]
```diff
@@ -228,48 +268,44 @@ func (r *ReconcileBroadcastJob) Reconcile(request reconcile.Request) (reconcile.
...
-			return err
-		})
-		if err != nil {
-			klog.Errorf("failed to update job %s, %v", job.Name, err)
+	return reconcile.Result{RequeueAfter: requeueAfter}, err
+}
+
+func (r *ReconcileBroadcastJob) updateJobStatus(request reconcile.Request, job *appsv1alpha1.BroadcastJob) error {
+	klog.Infof("Updating job %s status", job.Name)
+	fmt.Printf("---------%+v", job.Status)
```
Comment (Fei-Guo) [inline pkg/controller/broadcastjob/broadcastjob_controller.go]: Should we remove this line or print job.status in klog?

File: pkg/apis/apps/v1alpha1/broadcastjob_types.go [hunk 8]
```diff
@@ -123,8 +131,52 @@ type BroadcastJobStatus struct {
...
+	PhaseFailed BroadcastJobPhase = "failed"
+)
+
+// FailurePolicy indicates the behavior of the job, when failed pod is found.
+type FailurePolicy struct {
+	// Type indicates the type of FailurePolicyType.
+	Type FailurePolicyType
+
+	// RestartLimit specifies the number of retries before marking the pod failed.
+	RestartLimit int32
```
Comment (Fei-Guo) [inline pkg/apis/apps/v1alpha1/broadcastjob_types.go]: I think we should remove BackoffLimit *int32 from the API if we provide per pod restart limit. I don't see a case where people need both.

**Comment by Fei-Guo** [comment]
LG. Can you add some details in the change description? Especially because this change made significant API changes. A better document will make future query/search easier.

**Comment by xlgao-zju** [comment]
> Can you add some details in the change description?

Add the description in `docs/concepts/broadcastJob/README.md`?

**Comment by Fei-Guo** [comment]
> > Can you add some details in the change description?
> 
> Add the description in `docs/concepts/broadcastJob/README.md`?

I mean in this PR. We usually use the PR template in kruise/.github/PULL_REQUEST_TEMPLATE.md. I am not sure why this is not applied automatically in this PR.

**Comment by xlgao-zju** [comment]
@Fei-Guo I updated this PR according to the PR template.

**Pull request merged**